### PR TITLE
Drop recursive equation in Typing_env_extension.join

### PR DESCRIPTION
**Pull request by lukemaurer**
Fixes a corner case where we're joining `x : ty1` and `x : ty2` and it
turns out `ty1` and `ty2` are both aliases that canonicalize to `x`.
We were producing `x = x` in that case, which is invalid. The correct
behavior is to drop `x` since it's effectively unknown.

### CODE REVIEW

**Comment by lthls** [comment]
I'm not completely convinced that it's the best fix, but it looks correct.

I think we really need to avoid tautologic equations in the typing environment, but not necessarily in the extensions themselves; `Typing_env.add_equation` could simply drop any such equation.

**Review by lthls — APPROVED** [review]
I don't agree with the comment saying that `name = name` is an invalid equation, but it's true that it's unexpected and could cause errors later, so I'm fine with merging as it is.

File: middle_end/flambda2/types/env/typing_env_extension.rec.ml [hunk 1]
```diff
@@ -141,14 +142,34 @@ let meet env t1 t2 : _ Or_bottom.t =
     Ok (meet0 env t1 t2 [])
   with Bottom_meet -> Bottom
 
+let is_alias_of_name ty name =
```
Comment (mshinwell) [inline middle_end/flambda2/types/env/typing_env_extension.rec.ml:145]: This function should probably be in `Type_grammar`.

File: middle_end/flambda2/types/env/typing_env_extension.rec.ml [hunk 2]
```diff
@@ -141,14 +142,34 @@ let meet env t1 t2 : _ Or_bottom.t =
...
         | Some ty1, Some ty2 ->
           begin match Type_grammar.join env ty1 ty2 with
-          | Known ty -> Some ty
+          | Known ty ->
+            if is_alias_of_name ty name then
+              (* This is rare but not anomalous. It may mean that [ty1] and [ty2]
+                 are both alias types which canonicalize to [name], for instance.
+                 In any event, if the best type available for [name] is [= name],
+                 we effectively know nothing, so we drop [name]. ([name = name] is
+                 an invalid equation anyway.) *)
```
Comment (mshinwell) [inline middle_end/flambda2/types/env/typing_env_extension.rec.ml:166]: Let's reword this so it doesn't say "invalid" (c.f. @lthls 's comment)

**Review by mshinwell — CHANGES_REQUESTED** [review]

File: middle_end/flambda2/types/env/typing_env_extension.rec.ml (hunk 2 above)
Comment (lukemaurer) [reply middle_end/flambda2/types/env/typing_env_extension.rec.ml:166]: Good point. Maybe just say it would be rejected by `Typing_env.add_equation`?

Comment (mshinwell) [reply middle_end/flambda2/types/env/typing_env_extension.rec.ml:166]: Yeah that seems ok

**Review by mshinwell — APPROVED** [review]

**Pull request merged**